### ISSUE TITLE
[emerg] "if" directive is not allowed here in /etc/nginx/bots.d/blockbots.conf:40 (EasyEngine Users)

### ISSUE THREAD

**Issue by janiosarmento** [mitchellkrogza/nginx-ultimate-bad-bot-blocker#150]
Ubuntu 14.04, Nginx 1.10.3: after updating all the vhost files `nginx -t` reports broken configuration:

    [emerg] "if" directive is not allowed here in /etc/nginx/bots.d/blockbots.conf:40

**Comment by ScrewLooseDan** [comment]
On my Ubuntu/Nginx setup, I believe I had to manually put the includes inside the location block for the vhost as the script just put them at the end of the file.  So, my vhost files look something like:

```
server {
  listen 443 ssl;
  server_name example.com;

location / {
# Other Useful stuff

# Bad Bot Blocker
 include /etc/nginx/bots.d/ddos.conf;
 include /etc/nginx/bots.d/blockbots.conf;
 }
```
What do your vhost files look like?

**Comment by itoffshore** [comment]
line `395` of the setup script shows how the insertion point is calculated:

`line=$(find_includes $file include last server_ last location first )`

which means insert:
____________________

after the last `server_`

&

before the first `location` directive

&

after the last `include` (within this range `server_ <----> location`)
____________________

you can play around with the parameters passed to `find_includes()` to suit your particular case.

I will have a think about how to make these settings more accessible

**Comment by janiosarmento** [comment]
```

server {


    server_name janio.sarmento.org   www.janio.sarmento.org;

        # bot blocker was inserted here

    access_log /var/log/nginx/janio.sarmento.org.access.log rt_cache_redis;
    error_log /var/log/nginx/janio.sarmento.org.error.log;


    root /var/www/janio.sarmento.org/htdocs;

    index index.php index.html index.htm;

    include /var/www/janio.sarmento.org/conf/nginx/*.conf;

    include  common/redis-php7.conf;


    include common/wpcommon-php7.conf;
    include common/locations-php7.conf;
}
```

I have a location inside the `locations-php7.conf`.

I'm so sorry I've been bringing issues here. I tried to solve this issue by myself, but I'm not that good in Bash yet.

**Comment by itoffshore** [comment]
I think this may be a problem:

`include /var/www/janio.sarmento.org/conf/nginx/*.conf;`

(depending on which `*.conf` files these are)

you should keep your config files in somewhere like `/etc/nginx/conf.d` & **not** in your publicly accessible `www` directory

**Comment by itoffshore** [comment]
my vhost layout is : 

`includes` ....( with Bad Bot Blocker as the last includes)

`access_log`
`error_log`

`root /path/to/domain`

`location` blocks

**Comment by janiosarmento** [comment]
> you should keep your config files in somewhere like /etc/nginx/conf.d & not in your publicly accessible www directory

The domain public root is `/var/www/janio.sarmento.org/htdocs`; thus, my configuration files are safe, at the same time they are organized, easily included in a single `rsync` call for backup purposes.

I totally understand that your script might require some specific layout to vhosts file in order to work, but I can't accept that you assume my setup is not safe without actually understanding how it works.

I'll see what I can do in order to positively contribute with your project, since you kindly puts it open sourced for the entire community.

**Comment by janiosarmento** [comment]
I'm running `./setup-ngxblocker -e org -x` (being `org` each of my domain extensions) and so far no errors.

Probably it is my `default` vhost file that's causing issues (it listens the IP without any domain name attached, and redirects traffic to a "parked domain" generic page).

I'll do more testing and share my results.

**Comment by mitchellkrogza** [comment]
Thanks @janiosarmento  please let us know if you resolve it, often just one host, often it's the default, can cause untold problems

**Comment by mitchellkrogza** [comment]
@janiosarmento did you get your problems resolved?

**Comment by janiosarmento** [comment]
Yes and no.

Yes, I was able to install Bot Blocker in a number of domains.

No, it is not straightforward all the time, but I couldn't point out what's preventing the installation from working the first time.

In order to avoid my `default` vhost to create issues I `chattr +i` it before installing the filters.

Also, I have to manually remove DDoS filters manually because the rules conflict with previous `nginx.conf` rules (I read something about this possibility in documentation, I think).

By next monday I think I'll have something more to say (because I'll spend the entire weekend with my family, little to no chance to spend time with computer).

**Comment by itoffshore** [comment]
@janiosarmento - running `setup-ngxblocker` with the `-d` switch disables the insert of the DDOS configuration. 

All of the scripts can be run with `-h` to see their configuration options

**Comment by janiosarmento** [comment]
English is not my first language (you certainly noticed it). That's why, perhaps, I didn't interpret the `-h` information correctly.

Thanks for clarifying. :)

**Comment by mitchellkrogza** [comment]
Enjoy your weekend and Happy Easter and we do try our best to make sure everyone get's to run and enjoy the blocker.

**Comment by janiosarmento** [comment]
100% confirmed: the conflict occurs because of my `default` vhost.

In order to install the Bad Bot Blocker rules to my vhosts accordingly I had to:

```
for EXT in `find . -type f | sed -rn 's|.*/[^/]+\.([^/.]+)$|\1|p' | sort -u | grep -v conf`; do
    /usr/local/sbin/setup-ngxblocker -x -d -e ${EXT}
done
```

Of course, this only applies to my servers, which don't have any extension to vhost files (except for `default.conf`, that I ignore because it is the root of my problems).

It works for me, and it will work for anyone using EasyEngine.

**Comment by mitchellkrogza** [comment]
Thanks for the feedback and information @janiosarmento I am sure this will assist other users of EasyEngine. If possible could you post the contents of the default vhost here (removing any sensitive information of course) as it may help us to identify what causes the problem.

**Comment by janiosarmento** [comment]
There is nothing sensitive in my default vhost. Here it goes:

```
##
# You should look at the following URL's in order to grasp a solid understanding
# of Nginx configuration files in order to fully unleash the power of Nginx.
# http://wiki.nginx.org/Pitfalls
# http://wiki.nginx.org/QuickStart
# http://wiki.nginx.org/Configuration
#
# Generally, you will want to move this file somewhere, and start with a clean
# file but keep this around for reference. Or just disable in sites-enabled.
#
# Please see /usr/share/doc/nginx-doc/examples/ for more detailed examples.
##

# Default server configuration
#
server {
	listen 80 default_server;
	listen [::]:80 default_server;

	# SSL configuration
	#
	# listen 443 ssl default_server;
	# listen [::]:443 ssl default_server;
	#
	# Self signed certs generated by the ssl-cert package
	# Don't use them in a production server!
	# include snippets/snakeoil.conf;
	#
	# ssl_protocols TLSv1 TLSv1.1 TLSv1.2; # don’t use SSLv3 ref: POODLE
	# ssl_ciphers HIGH:!aNULL:!MD5;
	# ssl_prefer_server_ciphers on;

	root /var/www/html;

	# Add index.php to the list if you are using PHP
	index index.html index.htm index.nginx-debian.html;

	server_name _;

	location / {
		# First attempt to serve request as file, then
		# as directory, then fall back to displaying a 404.
		#try_files $uri $uri/ =404;

                # Not really, actually redirects all traffic to the "parked" page
                rewrite ^(.*) https://www.portofacil.net/suspended-page;
	}

	# pass the PHP scripts to FastCGI server listening on 127.0.0.1:9000
	#
	#location ~ \.php$ {
	#	include snippets/fastcgi-php.conf;
	#
	#	# With php5-cgi alone:
	#	fastcgi_pass 127.0.0.1:9000;
	#	# With php5-fpm:
	#	fastcgi_pass unix:/var/run/php5-fpm.sock;
	#}

	# deny access to .htaccess files, if Apache's document root
	# concurs with nginx's one
	#
	#location ~ /\.ht {
	#	deny all;
	#}
}


# Virtual Host configuration for example.com
#
# You can move that to a different file under sites-available/ and symlink that
# to sites-enabled/ to enable it.
#
#server {
#	listen 80;
#	listen [::]:80;
#
#	server_name example.com;
#
#	root /var/www/example.com;
#	index index.html;
#
#	location / {
#		try_files $uri $uri/ =404;
#	}
#}
```

**Comment by itoffshore** [comment]
This issue is avoided if people use a `vhost` file extension (as we recommend)

The correct way to add a default site configuration is for the main `nginx.conf` to have includes as follows:

```
http {
        ....
        # include vhosts after all http settings
        include /path/to/your/default.conf;
        include /etc/nginx/sites-enabled/*.vhost;

        # Bad Bot Blocker install script inserts here 
        # as last includes in http stanza
        include /etc/nginx/conf.d/globalblacklist.conf;
        include /etc/nginx/conf.d/botblocker-nginx-settings.conf;
}
```

if you already run an `nginx` jail or use `limit_conn` you probably also want to run `setup-ngxblocker` with the `-d` switch to disable the DDOS rule from `Bot Blocker`

**Comment by dutuka** [comment]
Hello,
I use EasyEngine on Ubuntu 14.04, Nginx 1.10.3 and have the same problem too.  

> [emerg] "if" directive is not allowed here in /etc/nginx/bots.d/blockbots.conf:40

I try the command  like janiosarmento did:

```
cd /etc/nginx/sites-available
for EXT in `find . -type f | sed -rn 's|.*/[^/]+\.([^/.]+)$|\1|p' | sort -u | grep -v conf`; do
    /usr/local/sbin/setup-ngxblocker -x -d -e ${EXT}
done
```
then
```
cd /usr/local/sbin/
sudo ./setup-ngxblocker -x
```
or 
```
sudo ./setup-ngxblocker -d
```

but it does not work for me.  Any advice how I can get the blocker worked? Thanks.

**Comment by itoffshore** [comment]
@dutuka - the solution is in **my message above** on **3rd April**

I think you are using a `default` file in your `sites-available` - if you want to use a `default` use a file structure like:

```
        # include vhosts after all http settings
        include /path/to/your/default.conf;
        include /etc/nginx/sites-enabled/*.vhost;
```

( do not keep your `default` file with your other `vhost` files so the above settings work in `nginx.conf`)

**Comment by dutuka** [comment]
**@itoffshore** -  thank you for your help, it works. After playing three time with the installation of the bot blocker, I found another way to get it worked, the good thing is the `default` file can stay in `sites-enabled` folder (its better because easyengine use this file structure) and I can run `setup-ngxblocker -x`, not `setup-ngxblocker -d`, so I can use the DDOS configuration.

I think my solution will work for other Easyengine users too, so this is what I did:

1. Go to /etc/nginx/ and backup sites-available, sites-enabled folders. Download them or use zip:
```
cd /etc/nginx/
zip -9 -r site-available.zip sites-available
zip -9 -r site-enabled.zip sites-enabled
```
2. Go to folder `/etc/nginx/sites-enabled/` and `/etc/nginx/sites-available/`, rename default file to `default.conf` and all vhost file from `mydomain.com to mydomain.com.vhost`

3. Go to [nginx-ultimate-bad-bot-blocker](https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker) and start with the installation from step 1 to step 5.
Note: at step 5 no need to run `sudo ./setup-ngxblocker -x -e conf`, 
only run `sudo ./setup-ngxblocker -x`

4. After step 5, all files in `/etc/nginx/sites-enabled/` are now empty, I dont know why but they are empty. Go to `/etc/nginx/`, delete `sites-enabled` folder and restore it from the zip file I backed up before with:
```
cd /etc/nginx/
unzip site-enabled.zip
```
5. Go to `sites-enabled` folder and insert manually to all vhost file in the `server {} block` following lines:
```
 	# Bad Bot Blocker
	include /etc/nginx/bots.d/ddos.conf; 
 	include /etc/nginx/bots.d/blockbots.conf;
```
For exsample: 

```
    server_name domain.com   www.domain.com;

 	# Bad Bot Blocker
	include /etc/nginx/bots.d/ddos.conf; 
 	include /etc/nginx/bots.d/blockbots.conf;
 

    access_log 
    error_log 
```
or copy them from folder `site-available` to `sites-enabled` may work too, I only have one site so I did it manually.

Note: no need to rename the default file and vhost files anymore. The ending `vhost` and `conf` is onlny needed for the installation from step 1 to step 5. 

6. Now test the nginx configuration with `sudo nginx -t`, if it shows:
```
nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
nginx: configuration file /etc/nginx/nginx.conf test is successful
```
then the bot blocker is successfully installed.

7. Continue the installation from step 6 to step 10. Done.
